scripts/build/build/__init__.py

```diff
@@ -0,0 +1,151 @@
+import logging
+import os
+import shutil
+
+from enum import Enum, auto
+from typing import Sequence
+
+from .targets import Platform, Board, Application
+from .factory import BuilderFactory, TargetRelations
+
+
+def CommaSeparate(items) -> str:
+  return ', '.join([x.ArgName for x in items])
+
+
+# Supported platforms/boards/apps for generation/compilation
+# note that not all combinations are supported, however command-line will
+# accept all combinations
+PLATFORMS = [x.ArgName for x in Platform]
+BOARDS = [x.ArgName for x in Board]
+APPLICATIONS = [x.ArgName for x in Application]
+
+
+class BuildSteps(Enum):
+  GENERATED = auto()
+
+
+class Context:
+  """Represents a grouped list of platform/board/app builders to use
+
+       to generate make/ninja instructions and to compile.
+    """
+
+  def __init__(self, repository_path, output_prefix):
+    self.builders = []
+    self.repository_path = repository_path
+    self.output_prefix = output_prefix
+    self.completed_steps = set()
+
+  def SetupBuilders(self, platforms: Sequence[Platform],
+                    boards: Sequence[Board],
+                    applications: Sequence[Application]):
+    """Configures internal builders for the given platform/board/app combionation.
+
+        Handles smart default selection, so that users only need to specify
+        part of platform/board/application information and the method tries
+        to automatically deduce the rest of the arguments.
+        """
+    if not platforms and not boards:
+      if applications:
+        platforms = set().union(*[
+            TargetRelations.PlatformsForApplication(app) for app in applications
+        ])
+      else:
+        # when nothing is specified, start with a default host build
+        platforms = [Platform.LINUX]
```

Comment:
This is not correct behavior on Mac.

Reply:
I have not added mac support yet. I plan to have 'host' as a platform instead (I don't see us cross-compiling anytime soon) however I would prefer to do that when I am actually testing on mac. For now adding a TODO